### PR TITLE
Update Crate Dependencies + posslibly fixed deserialization issue with items

### PR DESCRIPTION
**Pull request by ju6ge**
Hi,

i adjusted the code to be compatible with the newest version of reqwest and serde ;)

Kind regrads,
ju6ge

### CODE REVIEW

**Comment by ju6ge** [comment]
After some more debugging if have found some more inconsistencys with the behavior of the zotero api and fixed them. 
Possibly fixes #2

**Comment by ju6ge** [comment]
Hey,

it is a bit sad when projects die because the people involved do not have time anymore. Nobody should be on the hook for maintaining something if they have other things to do in life. 

So I like to ask if you would be willing to let me take over maintainership of this repositoy/crate? I have active use for it and would like to make sure that other people can too.

Thanks for creating it, it has saved me a lot of time.

Kind regards,
ju6ge

**Comment by joshkarlin** [comment]
yo @Eonm, are you willing to agree to the above?  I would also like to use this crate and if @ju6ge is willing to maintain it then seems like everyone will be happy

**Comment by ju6ge** [comment]
@joshkarlin I just wrote an email directly to the maintainer. Maybe email notifications where turned off. But the website in the profile is also offline. So maybe something happend? I hope not!

Is there a process in Github on how to deal with stuff like this? Or is the answer just hard fork it and create an new crate?

**Comment by joshkarlin** [comment]
yeah i hope all is well.

i'm not sure any github processes are available...  from [this discussion](https://github.com/orgs/community/discussions/13470), it looks like it's still an open issue.  i suspect hard fork would be the easiest and quickest

**Comment by Eonm** [comment]
Hello @ju6ge thank you for your work, I reviewed your pull request

My apologies,

I wasn't notified by GitHub and haven't checked my personal GitHub account in a while. I will review your pull request as soon as possible.

I didn't expect my library to be used by others. I need help to maintain this library, I will add you as maintainer of the project.

Best regards

File: zotero/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+edition = "2018"
+license = "MIT"
+description = "Communicate with Zotero API"
+repository = "https://github.com/Eonm/zotero"
+keywords = ["zotero", "digital-humanities", "citations", "zotero-api"]
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+reqwest = { version = "0.11.11", features = ["blocking", "json"] }
```
Comment (Eonm) [inline zotero/Cargo.toml:14]: Perhaps we should use the asynchronous version of reqwest in combination with tokio. Or do it in another pull request

File: zotero/src/data_structure/item/item_data/bill.rs [hunk 2]
```diff
@@ -1,11 +1,13 @@
...
-use serde::Serialize;
 use std::collections::HashMap;
+use serde::{Deserialize, Serialize};
+
+use zotero_derive::ItemCommon;
 
 /// A proposed piece of legislation.
-#[derive(Default, Deserialize, Serialize, Clone, Debug, Builder)]
+#[derive(Default, Deserialize, Serialize, Clone, Debug, Builder, ItemCommon)]
 // #[serde(rename_all(deserialize = "camelCase", serialize = "camelCase"))]
```
Comment (Eonm) [inline zotero/src/data_structure/item/item_data/bill.rs]: We should remove this line

```suggestion
```

File: zotero/src/data_structure/mod.rs [hunk 3]
```diff
@@ -1,11 +1,13 @@
...
 pub mod collection;
 pub mod item;
-mod shared_fields;
+pub mod shared_fields;
 
 /// Transform a data structure to a json object
-pub trait ToJson: serde::Serialize {
+pub trait ToJson: Serialize {
     fn to_json(&self) -> String {
         serde_json::to_string(&self).unwrap()
```
Comment (Eonm) [inline zotero/src/data_structure/mod.rs:12]: My initial implementation is a bit dirty.

We should remove unwrap from code base.

May be `use std::convert::TryInto;` is more idiomatic than a custom trait ?

File: zotero/src/data_structure/shared_fields.rs [hunk 4]
```diff
@@ -24,6 +23,32 @@ pub struct Link {
...
+
+    fn key(&self) -> &str;
+
+    fn has_tag(&self, tag: String) -> bool {
+        for t in self.tags() {
+            if tag == t.tag {
+                return true;
+            }
+        }
+        return false;
```
Comment (Eonm) [inline zotero/src/data_structure/shared_fields.rs]: Personally, I prefer a more functional style

```suggestion
      self.tags().iter().any(|t| t.tag == tag)
```

File: zotero/src/data_structure/item/mod.rs [hunk 5]
```diff
@@ -192,6 +202,289 @@ pub struct Item {
...
+        self.data.tags()
+    }
+
+    pub fn has_tag(&self, tag: &str) -> bool {
+        for t in self.tags() {
+            if tag == t.tag {
+                return true;
+            }
+        }
+        return false;
```
Comment (Eonm) [inline zotero/src/data_structure/item/mod.rs]:
```suggestion
self.tags().iter().any(|t| t.tag == tag)
```

File: zotero/src/data_structure/item/mod.rs [hunk 6]
```diff
@@ -204,17 +497,87 @@ pub struct Creator {
...
+impl Creator {
+    pub fn full_name(&self) -> String {
+        format!("{} {}", self.first_name, self.last_name)
+    }
+
+    pub fn short_name(&self) -> String {
+        if self.first_name.len() == 0 {
+            return format!("{}", self.last_name)
+        }
+        format!("{}. {}", self.first_name.chars().next().unwrap(), self.last_name)
```
Comment (Eonm) [inline zotero/src/data_structure/item/mod.rs]:
```suggestion
match self.first_name.chars().next() {
    Some(first_initial) => format !("{}. {}", first_initial, self.last_name),
    None => format !("{}", self.last_name),
}
```
This code doesn't work for some names: `Jean-Philippe` should be transformed into `J.-P.` or `Hanna Mary` into `H. M.`

Perhaps we should declare the surname as optional and set the value to None when deserializing an empty string?

File: zotero/src/data_structure/item/mod.rs [hunk 7]
```diff
@@ -204,17 +497,87 @@ pub struct Creator {
...
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+pub enum SizeOrBool {
```
Comment (Eonm) [inline zotero/src/data_structure/item/mod.rs:566]:
```suggestion
#[serde(untagged)]
pub enum SizeOrBool {
```

With `#[serde(untagged)]` serde will try to deserialize against each variant.

File: zotero/src/data_structure/item/mod.rs [hunk 8]
```diff
@@ -204,17 +497,87 @@ pub struct Creator {
...
+            Ok(v) => Ok(Some(SizeOrBool::Size(v)))
+        }
+    } else if s.is_null() {
+        Ok(None)
+    } else {
+        panic!("invalid type")
+    }
+}
+
+
```
Comment (Eonm) [inline zotero/src/data_structure/item/mod.rs]:
```suggestion
```

This code might not be necessary if untagged works (https://serde.rs/enum-representations.html#untagged)

File: zotero/src/data_structure/item/mod.rs [hunk 9]
```diff
@@ -192,6 +202,289 @@ pub struct Item {
...
+                ItemType::Note(d) => {
+                    &d.date_added
+                }
+        };
+        convert_zotero_date_str(&date_str)
+    }
+}
+
+fn convert_zotero_date_str(date_str: &str) -> DateTime<Local> {
+    let date_regex = Regex::new(r"(\d{4})-(\d{2})-(\d{2})").unwrap();
```
Comment (Eonm) [inline zotero/src/data_structure/item/mod.rs]: We should move this regex inside a lazy_static to avoid building the regex each time the function is called.

https://stackoverflow.com/questions/35169259/how-to-make-a-compiled-regexp-a-global-variable

File: zotero/src/data_structure/item/mod.rs [hunk 10]
```diff
@@ -192,6 +202,289 @@ pub struct Item {
...
+    let date_captures = date_regex.captures(date_str);
+
+    let formatter_regex = Regex::new(r"(\d{2})/(\d{4})").unwrap();
```
Comment (Eonm) [inline zotero/src/data_structure/item/mod.rs]: We should move this regex inside a lazy_static to avoid building the regex each time the function is called.

https://stackoverflow.com/questions/35169259/how-to-make-a-compiled-regexp-a-global-variable

File: zotero/Cargo.toml (hunk 1 above)
Comment (ju6ge) [reply zotero/Cargo.toml:14]: I was thinking the same … but I would argue that a transition to async code should be a seperate pull request!

File: zotero/src/data_structure/mod.rs (hunk 3 above)
Comment (ju6ge) [reply zotero/src/data_structure/mod.rs:12]: I am not sure we really need to explicitly provide an ToJson implementation. Serde handles this well, and a library user can do the conversion without any issue and handle it.

**Comment by ju6ge** [comment]
@Eonm great to hear from you. I will gladly help with the maintenance of this project ;) 

For now I did some quick review of the updates you suggested. For the rest I need some more time ;)

**Comment by Eonm** [comment]
@ju6ge Thanks for the changes.

Just one last thing, one of the unit tests doesn't pass. Once fixed, I'll merge the pull request

File: zotero/src/data_structure/item/mod.rs (hunk 8 above)
Comment (ju6ge) [reply zotero/src/data_structure/item/mod.rs]: You are right this makes it simpler, no custom deserialize function required.

File: zotero/src/data_structure/item/mod.rs (hunk 6 above)
Comment (ju6ge) [reply zotero/src/data_structure/item/mod.rs]: i changed the code to the suggest logic, a more though update would be needed to enable smarter handling of names, we should probably write some test cases to to ensure that it is handled correctly

**Comment by ju6ge** [comment]
So I updated the PR and cleaned up the code base with `cargo fmt` and `cargo clippy`. All test should be passing now. The next thing I will be working on is enabling async support. But that should be it's own pull request.
